### PR TITLE
maxSegmentsInQueue in CoordinatorDinamicConfig

### PR DESCRIPTION
**Pull request by dgolitsyn**
maxSegmentsInQueue parameter added to CoordinatorDinamicConfig and used in LoadRule to improve segments loading and replication speed. Adjusting this config parameter might be useful if there are "slow" nodes in the cluster to prevent huge backlog in loading queue for such nodes. It could be more important to load segment to not optimal (in sense of BalancerStrategy) server than waiting for a long time until backlog is processed.
Default value does not change behavior.

### CODE REVIEW

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 1]
```diff
@@ -109,7 +113,8 @@ public CoordinatorDynamicConfig(
       int balancerComputeThreads,
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Could this constructor be private and it's usages are replaced with builder?

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 2]
```diff
@@ -94,6 +96,8 @@ public CoordinatorDynamicConfig(
                                    ? current.getKillDataSourceWhitelist()
                                    : parseKillDataSourceWhitelist(killDataSourceWhitelist);
 
+    this.maxSegmentsInQueue = maxSegmentsInQueue == null ? current.getMaxSegmentsInQueue() : maxSegmentsInQueue;
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: From where the `current.getMaxSegmentsInQueue()` value will come?

For backwards compatibility, I think maxSegmentsInQueue=0 should mean unlimited.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 3]
```diff
@@ -43,6 +43,7 @@
   private final boolean emitBalancingStats;
   private final boolean killAllDataSources;
   private final Set<String> killDataSourceWhitelist;
+  private final int maxSegmentsInQueue;
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Please document this property in code, e. g. it's not clear what queue? per-node queue, global queue? Add refs. Consider renaming to `maxSegmentsInNodeQueue` or something like that.

Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Also add docs to `docs/content/configuration/coordinator.md`

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 4]
```diff
@@ -60,7 +61,8 @@ public CoordinatorDynamicConfig(
       // coordinator console can not send array of strings in the update request.
       // See https://github.com/druid-io/druid/issues/3055
       @JsonProperty("killDataSourceWhitelist") Object killDataSourceWhitelist,
-      @JsonProperty("killAllDataSources") Boolean killAllDataSources
+      @JsonProperty("killAllDataSources") Boolean killAllDataSources,
+      @JsonProperty("maxSegmentsInQueue") Integer maxSegmentsInQueue
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Could you please add deserialization test, showing that deserialization of an old config creates an object with unbounded queue?

File: server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java [hunk 5]
```diff
@@ -64,12 +65,16 @@ public CoordinatorStats run(DruidCoordinator coordinator, DruidCoordinatorRuntim
...
       final MinMaxPriorityQueue<ServerHolder> serverQueue = params.getDruidCluster().getHistoricalsByTier(tier);
+      final int maxSegmentsInQueue = params.getCoordinatorDynamicConfig().getMaxSegmentsInQueue();
       if (serverQueue == null) {
         log.makeAlert("Tier[%s] has no servers! Check your cluster configuration!", tier).emit();
         continue;
       }
 
-      final List<ServerHolder> serverHolderList = Lists.newArrayList(serverQueue);
+      final List<ServerHolder> serverHolderList = serverQueue.stream()
+                                                             .filter(s -> s.getNumberOfSegmentsInQueue() < maxSegmentsInQueue)
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java]: This line is longer than 120 columns

**Comment by leventov** [comment]
Also this PR fails some tests, not flaky, related to this change.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 2 above)
Comment (leiwangx) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: does it make sense for unlimited to mean anything that's < 1.

File: server/src/test/java/io/druid/server/coordinator/rules/LoadRuleTest.java [hunk 6]
```diff
@@ -524,4 +520,115 @@ public boolean appliesTo(Interval interval, DateTime referenceTimestamp)
...
+    exec.shutdown();
+  }
+
+  private DataSegment createDataSegment(String dataSource)
+  {
+    return new DataSegment(
+        dataSource,
+        new Interval("0/3000"),
+        new DateTime().toString(),
+        Maps.<String, Object>newHashMap(),
```
Comment (leiwangx) [inline server/src/test/java/io/druid/server/coordinator/rules/LoadRuleTest.java]: Line 626:628, collection type can be auto inferred.

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 7]
```diff
@@ -297,10 +314,11 @@ public int hashCode()
     private int balancerComputeThreads;
     private Set<String> killDataSourceWhitelist;
     private boolean killAllDataSources;
+    private int maxSegmentsInQueue = Integer.MAX_VALUE;
```
Comment (leiwangx) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: do we need to init `maxSegmentsInQueue` to `Integer.MAX_VALUE`?

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 2 above)
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: @leiwangx I think widely used convention is just 0=unlimited. Negative values could be result of typo and should be rejected IMO.

File: server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java [hunk 8]
```diff
@@ -62,17 +63,24 @@ public CoordinatorStats run(DruidCoordinator coordinator, DruidCoordinatorRuntim
...
+
       if (serverQueue == null) {
         log.makeAlert("Tier[%s] has no servers! Check your cluster configuration!", tier).emit();
         continue;
       }
 
+      final int maxSegmentsInNodeLoadingQueue = params.getCoordinatorDynamicConfig()
+                                                      .getMaxSegmentsInNodeLoadingQueue();
+
+      Predicate<ServerHolder> serverHolderPredicate = (maxSegmentsInNodeLoadingQueue > 0) ?
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java]: Please use if-else

File: server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java [hunk 9]
```diff
@@ -62,17 +63,24 @@ public CoordinatorStats run(DruidCoordinator coordinator, DruidCoordinatorRuntim
...
+          s -> s != null && s.getNumberOfSegmentsInQueue() < maxSegmentsInNodeLoadingQueue :
+          s -> s != null;
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java]: Use `Objects::nonNull`

File: docs/content/configuration/coordinator.md [hunk 10]
```diff
@@ -93,6 +93,7 @@ Issuing a GET request at the same URL will return the spec that is currently in
 |`emitBalancingStats`|Boolean flag for whether or not we should emit balancing stats. This is an expensive operation.|false|
 |`killDataSourceWhitelist`|List of dataSources for which kill tasks are sent if property `druid.coordinator.kill.on` is true.|none|
 |`killAllDataSources`|Send kill tasks for ALL dataSources if property `druid.coordinator.kill.on` is true. If this is set to true then `killDataSourceWhitelist` must not be specified or be empty list.|false|
+|`maxSegmentsInNodeLoadingQueue`|The maximum number of segments that could be queued for loading to any given server. Default value is 0 (loading queue is unbounded) |0|
```
Comment (leventov) [inline docs/content/configuration/coordinator.md]: Document what this "maximum number segments" really means for the cluster. The current docs merely repeats the config name and doesn't actually explain it. The doc doesn't provide any hint to what number a user may want to configure `maxSegmentsInNodeLoadingQueue`.

**Pull request merged**